### ISSUE TITLE
Add a way to tell N805 that a class definition is a metaclass

### ISSUE THREAD

**Issue by flying-sheep** [astral-sh/ruff#20225]
### Summary

consider the following:

```py
class BaseMeta(type): pass
class MyMeta(BaseMeta):
    def thing(cls):  # <- N805
        assert isinstance(cls, type)


# runtime tests
import pytest
class MyCls(metaclass=MyMeta): pass

MyCls.thing()
with pytest.raises(AttributeError):
    MyCls().thing()
```

Ruff will flag `thing` as a [`N805`](https://docs.astral.sh/ruff/rules/invalid-first-argument-name-for-method/) violation, thinking that `MyMeta` is a regular class and should have `self` as first arg name.

The only way around it is adding `type` to `MyMeta`’s base classes, which is probably fine, but it is a code change just to make Ruff understand something.

Would be great if there could be way to make Ruff understand which doesn’t require code changes, like a `lint.pep8-naming.metaclasses` option (or something smarter / more general if you can come up with something)

**Comment by ntBre** [comment]
I'm not seeing N805 on this example in the [playground](https://play.ruff.rs/1b521c60-98ec-4318-b367-8511c47bc205) or in a local CLI test. I can reproduce this if `BaseMeta` is defined in a different file, though, in which case this would be a multi-file analysis limitation.

**Comment by flying-sheep** [comment]
You’re right! What’s the plan with multi-file analysis?

And what about morecomplex siutations, like if the meta base class is imported from a 3rd party package? `ty` has to do it like that, might Ruff in the future be able to be run in an environment-enabled mode where it leverages `ty`’s global analysis?

**Comment by ntBre** [comment]
Yeah, we've grouped the multi-file analysis with type-inference in our GitHub labels since ty handles them both. The hope is that we can reuse some of that infrastructure in the future!

**Comment by flying-sheep** [comment]
OK! Then I’ll just `noqa` it and close this, I assume it’ll come at some point automatically (no reason not to make things like this automatically multi-file once the infrastructure exists)
